lib/df-detection.c

```diff
@@ -44,27 +44,29 @@ void add_address(const void *addr, size_t len, unsigned long caller)
 }
 void start_system_call(long syscall)
 {
-	current->syscall_num = syscall;
-	current->addresses = (struct df_address_range *)kmalloc_array(
-	    DF_INIT_SIZE, sizeof(struct df_address_range), GFP_KERNEL);
-	current->sz = current->addresses ? DF_INIT_SIZE : 0;
-	current->num_read = 0;
-	current->pairs = (struct df_pair *)kmalloc_array(
-	    DF_INIT_SIZE, sizeof(struct df_pair), GFP_KERNEL);
-	current->df_size = current->pairs ? DF_INIT_SIZE : 0;
-	current->df_index = 0;
+	if (current->df_enable) {
+		current->syscall_num = syscall;
+		current->addresses = (struct df_address_range *)kmalloc_array(
+		    DF_INIT_SIZE, sizeof(struct df_address_range), GFP_KERNEL);
+		current->sz = current->addresses ? DF_INIT_SIZE : 0;
+		current->num_read = 0;
+		current->pairs = (struct df_pair *)kmalloc_array(
+		    DF_INIT_SIZE, sizeof(struct df_pair), GFP_KERNEL);
+		current->df_size = current->pairs ? DF_INIT_SIZE : 0;
+		current->df_index = 0;
+	}
 }
 void end_system_call(void)
 {
-	if (current->pairs != NULL) {
+	if (current->df_enable && current->pairs != NULL) {
```

Comment:
Checking for the same condition twice often flags that the code can be simplified.
Early return is much more preferable here.